File: 102_binary_tree_level_order_traversal/Solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def create_tree_using_bfs(nums: list) -> TreeNode:
    arr_len = len(nums)
    node_list = []

    for i in range(arr_len):
        new_node = TreeNode(nums[i])
        node_list.append(new_node)
        if i == 0: continue
        if i % 2 == 1: node_list[(i - 1) // 2].left = new_node
        else: node_list[(i - 1) // 2].right = new_node

    return node_list[0]

class Solution:
    def levelOrder(self, root: TreeNode) -> list:
        if not root: return []

        result, queue = [], [root]
        while queue:
            tmp_res = []
            for _ in range(len(queue)):
                cur = queue[0]
                queue = queue[1:]
                tmp_res.append(cur.val)
                if cur.left: queue.append(cur.left)
                if cur.right: queue.append(cur.right)
            result.append(tmp_res)
        return result

if __name__ == '__main__':
    nums = [i for i in range(14)]
    root = create_tree_using_bfs(nums)
    for tmp in Solution().levelOrder(root):
        print(tmp)

File: 102_binary_tree_level_order_traversal/Solution.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {};
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public static TreeNode createTreeUsingBFS(int[] nums) {
        TreeNode[] nodeArr = new TreeNode[nums.length];
        for (int i = 0; i < nums.length; i++) {
            nodeArr[i] = new TreeNode(nums[i]);
            if (i == 0) continue;
            if (i % 2 == 1) nodeArr[(i - 1) / 2].left = nodeArr[i];
            else nodeArr[(i - 1) / 2].right = nodeArr[i];
        }

        return nodeArr[0];
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) return result;

        Deque<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            List<Integer> tmpRes = new ArrayList<>();
            for (int i = queue.size(); i > 0; i--) {
                TreeNode curNode = queue.pollFirst();
                tmpRes.add(curNode.val);
                if (curNode.left != null) queue.add(curNode.left);
                if (curNode.right != null) queue.add(curNode.right);
            }
            result.add(tmpRes);
        }

        return result;
    }

    public static void main(String[] args) {
        TreeNode root = createTreeUsingBFS(new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10});
        List<List<Integer>> result = new Solution().levelOrder(root);
        for (List<Integer> tmp : result) {
            System.out.println(tmp);
        }
    }
}
